### PR TITLE
L2 support for Indexer Agent

### LINKED ISSUE #675: Failed to collect receipts in exchange for query fee voucher: invalid BigNumber value
After I close an allocation or reallocate, I see the following error in my indexer logs:

```
{
  "level": 50,
  "time": 1688402149262,
  "pid": 1,
  "hostname": "indexer-agent-0",
  "name": "IndexerAgent",
  "component": "AllocationReceiptCollector",
  "allocation": "0x1F9A8C681D25E3Ec2a5934a10E0DA86815493F0f",
  "function": "obtainReceiptsVoucher()",
  "err": {
    "type": "IndexerError",
    "message": "Failed to collect receipts in exchange for query fee voucher",
    "stack": "IndexerError: Failed to collect receipts in exchange for query fee voucher\n    at indexerError (/opt/indexer/packages/indexer-common/dist/errors.js:171:12)\n    at AllocationReceiptCollector.<anonymous> (/opt/indexer/packages/indexer-common/dist/allocations/query-fees.js:322:147)\n    at Generator.throw (<anonymous>)\n    at rejected (/opt/indexer/packages/indexer-common/dist/allocations/query-fees.js:6:65)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
    "code": "IE054",
    "explanation": "https://github.com/graphprotocol/indexer/blob/main/docs/errors.md#ie054",
    "cause": {
      "type": "Error",
      "message": "invalid BigNumber value (argument=\"value\", value=undefined, code=INVALID_ARGUMENT, version=bignumber/5.7.0)",
      "stack": "Error: invalid BigNumber value (argument=\"value\", value=undefined, code=INVALID_ARGUMENT, version=bignumber/5.7.0)\n    at Logger.makeError (/opt/indexer/node_modules/@ethersproject/logger/lib/index.js:238:21)\n    at Logger.throwError (/opt/indexer/node_modules/@ethersproject/logger/lib/index.js:247:20)\n    at Logger.throwArgumentError (/opt/indexer/node_modules/@ethersproject/logger/lib/index.js:250:21)\n    at Function.BigNumber.from (/opt/indexer/node_modules/@ethersproject/bignumber/lib/bignumber.js:239:23)\n    at toBN (/opt/indexer/node_modules/@ethersproject/bignumber/lib/bignumber.js:292:25)\n    at BigNumber.add (/opt/indexer/node_modules/@ethersproject/bignumber/lib/bignumber.js:57:43)\n    at AllocationReceiptCollector.<anonymous> (/opt/indexer/packages/indexer-common/dist/allocations/query-fees.js:305:26)\n    at Generator.next (<anonymous>)\n    at fulfilled (/opt/indexer/packages/indexer-common/dist/allocations/query-fees.js:5:58)\n    at runMicrotasks (<anonymous>)",
      "reason": "invalid BigNumber value",
      "code": "INVALID_ARGUMENT",
      "argument": "value"
    }
  },
  "msg": "Failed to collect receipts in exchange for an on-chain query fee voucher"
}
```

version:

```
root@indexer-agent-0:/opt/indexer/packages/indexer-agent# cat package.json
{
  "name": "@graphprotocol/indexer-agent",
  "version": "0.20.17",
...
}
```

### PR DESCRIPTION
**Pull request by tilacog**
Adds the capacity for the Agent to interact with multiple protocol networks (Blockchains that host The Graph Protocol) by running a single instance of the `indexer-agent` program.

TODO:
- [x] CHANGELOG
- [x] Write a migration guide
- [x] Provide sample network configuration YAML files
- [x] Fix `indexer-cli` tests
- [x] Write a `down` migration path
- [ ] Review if the `GraphNode` class needs an array of index node ids, in light of this Graph Node PR:
  -  https://github.com/graphprotocol/graph-node/pull/4695
  
  
Fixes #675

### CODE REVIEW

**Comment by tilacog** [comment]
Force pushed to rebase on top of latest `main`

File: packages/indexer-common/src/indexer-management/models/indexing-rule.ts
Comment (hopeyen) [inline packages/indexer-common/src/indexer-management/models/indexing-rule.ts:158]: is this changed to false because we allow identifier to be the same across networks? if so, can there be a check where identifier+network is still unique?

File: packages/indexer-common/src/operator.ts
Comment (hopeyen) [inline packages/indexer-common/src/operator.ts:74]: Does it make sense to generalize some of the queries made to `indexerManagementClient` field so that not every operator own a client? or if there exists a query that's network specific and forces operator to require indexer management access?

File: packages/indexer-service/src/commands/start.ts
Comment (hopeyen) [inline packages/indexer-service/src/commands/start.ts:409]: do you see situations where multiple network is provided to the indexer service, such as multiple gateway endpoints?

**Review by hopeyen — COMMENTED** [review]
The changes are looking very impressive, only functional people can make some of these updates! I skimmed through briefly and left some minor questions. will take a more careful look on a later date 💯 💯 💯

**Comment by tilacog** [comment]
Force pushed to rebase on top of latest `main`

File: packages/indexer-agent/src/commands/common-options.ts
Comment (tilacog) [reply packages/indexer-agent/src/commands/common-options.ts:118]: To use network specificaiton files (for L2/multinetwork operation), you need to execute `indexer start-multiple` instead of just `start`. The new subcommand has a different argument parsing tree, in which you can specify the `--network-specifications-directory` option, that will read all network specification YAML files under a directory.

The `--config-file` option can be used in both subcommands, but the `indexerOptions.mnemonic` is specific to network specification files, thus, will only work if you use `start-multiple`. If you want to use a config file with `start`, then that option is named just `mnemonic`.

File: packages/indexer-cli/src/commands/indexer/rules/get.ts
Comment (tilacog) [reply packages/indexer-cli/src/commands/indexer/rules/get.ts:63]: Since the protocol network is now part of the `IndexingRuleIdentifier` type, we need it to identify (and lookup) a single indexing rule uniquely.

File: packages/indexer-agent/src/agent.ts
Comment (tilacog) [reply packages/indexer-agent/src/agent.ts:161]: This check validates that an array of Networks is aligned with another array of Operators, in the sense that pairs at equal positions are configured with the same network specification.

File: packages/indexer-common/src/indexer-management/models/indexing-rule.ts
Comment (tilacog) [reply packages/indexer-common/src/indexer-management/models/indexing-rule.ts:158]: That's correct. This change is needed specifically for the `global` rule, which will now exist 1:1 with protocol networks.
So we have a global rule for `mainnet` and another global rule for `arbitrum-one`.
The check is done by the new primary key constraint (`identifier+protocolNetwork`). I believe Sequelize understand this too and may enforce this at application level.

File: packages/indexer-common/src/operator.ts
Comment (tilacog) [reply packages/indexer-common/src/operator.ts:74]: I really want to decouple the `IndexerManagementClient` from the `Operator` class. 
I ended up with a dependency cycle between them and the `Network` class, which I couldn't resolve.

Ideally, the IndexerManagementClient should be a separate service, but I believe this will be a whole new endeavour.

File: packages/indexer-service/src/commands/start.ts
Comment (tilacog) [reply packages/indexer-service/src/commands/start.ts:409]: Absolutely! 
For now, users need to run one `indexer-service` instance per network. However, I believe a single instance could potentially handle multiple networks.

File: packages/indexer-agent/src/agent.ts
Comment (tilacog) [reply packages/indexer-agent/src/agent.ts:224]: Yes! Nice catch!
I was rewriting this from the previous version of the code, part by part, and didn't notice it could be wrapped into a single operation

Comment (tilacog) [reply packages/indexer-agent/src/agent.ts]: Fixed in 1c006ad7

File: packages/indexer-cli/src/commands/indexer/rules/get.ts
Comment (hopeyen) [reply packages/indexer-cli/src/commands/indexer/rules/get.ts:63]: Does it really make sense to require protocol network whenever indexer-cli is used to get an indexing rule? Then a user might need to always specify "goerli" even though they only indexes "goerli" subgraph right? 
maybe it should only be required if the result isn't unique already or just print out the results that matches identifier if protocol network isn't specified?

File: packages/indexer-agent/src/agent.ts
Comment (hopeyen) [reply packages/indexer-agent/src/agent.ts]: It seems that Networks and Operators both came from mapping on `NetworkSpecification`, so I would like to understand the flow to separately derive Networks and Operators from `NetworkSpecification`, validate their positions as two arrays, then zip into "NetworkAndOperator". Is it not possible to simply validate NetworkSpecification for its Network and Operator info directly?

**Comment by tilacog** [comment]
Force pushed to squash commits

File: packages/indexer-cli/src/command-helpers.ts
Comment (fordN) [inline packages/indexer-cli/src/command-helpers.ts:240]: Would it make sense to merge `extractProtocolNetworkOption()` together with `validateNetworkIdentifier()`?

File: packages/indexer-cli/src/command-helpers.ts
Comment (fordN) [inline packages/indexer-cli/src/command-helpers.ts:247]: Worth having separate functions? Couldn't one just use extractProtocolNetworkOption with required=true?

File: packages/indexer-cli/src/commands/indexer/rules/get.ts
Comment (fordN) [reply packages/indexer-cli/src/commands/indexer/rules/get.ts:63]: It would also be nice if a user could set a default protocolNetwork value during a indexer-cli worksession and from then on it would just use that value for all commands so you don't have to type it on everything?

File: packages/indexer-service/src/commands/start.ts
Comment (fordN) [inline packages/indexer-service/src/commands/start.ts:42]: ?

**Review by fordN — CHANGES_REQUESTED** [review]

**Comment by fordN** [comment]
Have we been able to observe successful collecting of fees in multi network mode?

File: packages/indexer-agent/src/commands/common-options.ts
Comment (tilacog) [reply packages/indexer-agent/src/commands/common-options.ts]: Fixed in e6f1e5d5

File: packages/indexer-agent/src/__tests__/indexer.ts
Comment (tilacog) [reply packages/indexer-agent/src/__tests__/indexer.ts:162]: No, this is just a randomly generated address to an empty account on Goerli.

File: packages/indexer-agent/src/agent.ts
Comment (tilacog) [reply packages/indexer-agent/src/agent.ts]: Reworded it in 7f62f889

File: packages/indexer-agent/src/agent.ts
Comment (tilacog) [reply packages/indexer-agent/src/agent.ts:252]: It spawns and coordinates all the Eventuals required for the reconciliation loop. Maybe we could name it something like `buildReconciliationTree`?

File: packages/indexer-cli/src/commands/indexer/rules/get.ts
Comment (tilacog) [reply packages/indexer-cli/src/commands/indexer/rules/get.ts:63]: I think it's possible. Could the `indexer connect` config file hold that information? If so, would that be a good place to store the default value?

File: packages/indexer-agent/src/agent.ts
Comment (tilacog) [reply packages/indexer-agent/src/agent.ts]: I agree. That sounds like a more natural way to do it. The current design may have evolved organically as operators were added later.
Still, I think the Agent constructor could be simplified to receive an array of NetworkSpecifications and the IndexerManagementClient instance, as per your suggestion.

Comment (tilacog) [reply packages/indexer-agent/src/agent.ts]: I would love to address this as a separate issue, if possible, to not block the release.

File: packages/indexer-agent/src/db/migrations/11-add-protocol-network-field.ts
Comment (tilacog) [reply packages/indexer-agent/src/db/migrations/11-add-protocol-network-field.ts]: Fixed in 8a511172

File: packages/indexer-agent/src/agent.ts
Comment (tilacog) [reply packages/indexer-agent/src/agent.ts:1136]: Could you please clarify what parallel allocation means in this context?
Also, I noticed that if we change from `filter` to `find`, we need to update the signatures for the `Operator.closeEligibleAllocations` and `Agent.identifyExpiringAllocations` methods. (Is this the parallel allocations you are referring to?)
Last, could we move this to a new issue to unblock the release?

File: packages/indexer-cli/src/command-helpers.ts
Comment (tilacog) [reply packages/indexer-cli/src/command-helpers.ts:247]: While it might appear redundant, `extractProtocolNetwork` has a nullable return type, whereas `requireProtocolNetwork` has a stricter return type. This difference significantly aids the code where they are used, as it eliminates the need for conditional checks to account for missing values.

File: packages/indexer-service/src/commands/start.ts
Comment (tilacog) [reply packages/indexer-service/src/commands/start.ts]: I think that comment is outdated. It was made when we were considering multiplying the startup arguments at the CLI level, but we decided to switch to network specification files instead. So, that comment is no longer relevant.


Removed it in 6cfbfd57

File: packages/indexer-agent/src/agent.ts
Comment (tilacog) [reply packages/indexer-agent/src/agent.ts:1219]: Opened an issue for this https://github.com/graphprotocol/indexer/issues/706

File: packages/indexer-agent/src/cost.ts
Comment (tilacog) [reply packages/indexer-agent/src/cost.ts:61]: Addressed in #707

File: packages/indexer-agent/src/commands/start.ts
Comment (tilacog) [reply packages/indexer-agent/src/commands/start.ts]: This comment was outdated.
Removed in 04bf7afd

File: packages/indexer-agent/src/agent.ts
Comment (tilacog) [reply packages/indexer-agent/src/agent.ts]: Fixed in dd9648ed

File: packages/indexer-agent/src/agent.ts
Comment (fordN) [reply packages/indexer-agent/src/agent.ts:257]: For consistency right now let's use `component`. Though I agree that `context` may be a better label.

File: packages/indexer-agent/src/agent.ts
Comment (fordN) [reply packages/indexer-agent/src/agent.ts:1135]: Yep, I don't think this is important to get in with this PR.  Let's create an issue to track?

File: packages/indexer-agent/src/agent.ts
Comment (tilacog) [reply packages/indexer-agent/src/agent.ts]: Resolved in 0e258aa1

File: packages/indexer-agent/src/agent.ts
Comment (tilacog) [reply packages/indexer-agent/src/agent.ts]: Referenced in a new issue https://github.com/graphprotocol/indexer/issues/709

File: packages/indexer-agent/src/agent.ts
Comment (tilacog) [reply packages/indexer-agent/src/agent.ts:1012]: Referenced it in a new issue: https://github.com/graphprotocol/indexer/issues/710

**Review by fordN — APPROVED** [review]
Looks good! 👍 🚀

**Pull request merged**
